modules/phase_field/src/materials/ComputeGBMisorientationType.C

```diff
@@ -0,0 +1,315 @@
+{
+  quat_angle[grain_id][0] = cos(euler_angle[grain_id][1] / 2.) *
+                            cos((euler_angle[grain_id][0] + euler_angle[grain_id][2]) / 2.);
+  quat_angle[grain_id][1] = sin(euler_angle[grain_id][1] / 2.) *
+                            cos((euler_angle[grain_id][0] - euler_angle[grain_id][2]) / 2.);
+  quat_angle[grain_id][2] = sin(euler_angle[grain_id][1] / 2.) *
+                            sin((euler_angle[grain_id][0] - euler_angle[grain_id][2]) / 2.);
+  quat_angle[grain_id][3] = cos(euler_angle[grain_id][1] / 2.) *
+                            sin((euler_angle[grain_id][0] + euler_angle[grain_id][2]) / 2.);
+}
+
+// Function to multiply quaternions and update
+void
+ComputeGBMisorientationType::getQuaternionProduct(const double qi[4],
+                                                  const double qj[4],
+                                                  double q[4])
+{
+  q[0] = qi[0] * qj[0] - qi[1] * qj[1] - qi[2] * qj[2] - qi[3] * qj[3];
+  q[1] = qi[0] * qj[1] + qi[1] * qj[0] + qi[2] * qj[3] - qi[3] * qj[2];
+  q[2] = qi[0] * qj[2] - qi[1] * qj[3] + qi[2] * qj[0] + qi[3] * qj[1];
+  q[3] = qi[0] * qj[3] + qi[1] * qj[2] - qi[2] * qj[1] + qi[3] * qj[0];
+}
+
+// Function to return the misorientation of two quaternions
+double
+ComputeGBMisorientationType::getMisorientationFromQuaternion(const double qi[4], const double qj[4])
+{
+  double miso0, misom = MY_2PI;
+  double q[4], qib[4], qjb[4], qmin[4] = {0, 0, 0, 0};
+
+  for (int o1 = 0; o1 < o_sym; o1++)
+  {
+    for (int o2 = 0; o2 < o_sym; o2++)
+    {
+      getQuaternionProduct(sym_quat[o1], qi, qib);
+      getQuaternionProduct(sym_quat[o2], qj, qjb);
+
+      // j-grain conjugate quaternion
+      qjb[1] = -qjb[1];
+      qjb[2] = -qjb[2];
+      qjb[3] = -qjb[3];
+      getQuaternionProduct(qib, qjb, q);
+      miso0 = round(2 * acos(q[0]) * 1e5) / 1e5;
+
+      if (miso0 > MY_PI)
+        miso0 = miso0 - MY_2PI;
+      if (fabs(miso0) < misom)
+      {
+        misom = fabs(miso0);
+        qmin[0] = q[0];
+        qmin[1] = q[1];
+        qmin[2] = q[2];
+        qmin[3] = q[3];
+      }
+    }
+  }
+
+  miso0 = 2 * acos(qmin[0]);
+  if (miso0 > MY_PI)
+    miso0 = miso0 - MY_2PI;
+
+  return fabs(miso0);
+}
+
+// Get the Misorientation angle
+void
+ComputeGBMisorientationType::getMisorientationAngles()
+{
+  // Initialize symmetry operator as quaternion vectors
+  defineSymmetryOperator(&sym_quat);
+  int grain_num = _ebsd_reader.getGrainNum();
+  euler_angle.assign(grain_num, std::vector<double>(3));
+  quat_angle.assign(grain_num, std::vector<double>(4));
+
+  // Get Euler Angle of Orientation
+  for (int i = 0; i < grain_num; i++)
+  {
+    const EulerAngles & d = _ebsd_reader.getEulerAngles(i);
+    auto grain_id = _ebsd_reader.getFeatureID(i);
+    euler_angle[grain_id][0] = d.phi1 / 180 * MY_PI;
+    euler_angle[grain_id][1] = d.Phi / 180 * MY_PI;
+    euler_angle[grain_id][2] = d.phi2 / 180 * MY_PI;
+    eulerOrientationToQuaternion(grain_id);
```

Comment:
```suggestion
    quat_angle[grain_id] = euler_angle[grain_id].toQuaternion()
```

and change the type of `quat_angle` accordingly.